### ISSUE TITLE
Second pass for the lib

### ISSUE THREAD

**Issue by pkrawat1** [aviabird/gringotts#2]
Abstract/extract following modules.

Checklist 
- [x] 1. Response struct creation for all the gateways to return a common API. assigned to @SagarKarwande #1  
- [x] 2. Request module for wrapping all the requests 
- [ ] 3. SSL requests also which need a certificate in the application to make a call(need to research on this more).

**Comment by oyeb** [comment]
## On task 3

I honestly don't know what ActiveMerchant's [cacert.pm](https://github.com/activemerchant/active_merchant/blob/master/lib/certs/cacert.pem) is doing in their repo (seems to be [used](https://github.com/activemerchant/active_merchant/blob/2855e9a2d7dd7ad693d37efcd0638d8cde08e957/lib/active_merchant/connection.rb#L101) here reasonably). This is [Mozilla's CA bundle](https://www.mozilla.org/en-US/about/governance/policies/security-group/certs/), which is available on literally every OS via the system's `ssl` module. Browsers "trust" CAs that feature in this list. All good HTTP(S) client libs will use the system's certificates to make `https` requests right out of the box. [Like python's `requests` library](http://docs.python-requests.org/en/master/user/advanced/#ssl-cert-verification).

### Client side SSL

This answer on [SO helps](https://superuser.com/questions/353038/what-is-the-difference-between-client-side-and-server-side-certificates).
Some gateways require/allow a client (ie, a merchant) to authenticate itself using a Client-Side SSL Certificate _instead_ of using an API token or other popular methods. For such gateways, we need to send this client cert in the request.

[`HTTPoison`](https://hexdocs.pm/httpoison/HTTPoison.html#summary) uses [`hackney`](https://github.com/benoitc/hackney) and any SSL options passed to `HTTPoison`/`hackney` fall through to Erlang's SSL module. So, to send a client cert via `HTTPoison`, just pass the required SSL options `cert` or `certfile` as [documented](http://erlang.org/doc/man/ssl.html).

### Server side SSL

A gateway who's SSL cert is not trusted by Mozilla's CA ~~shouldn't~~ won't exist, so we'll never have to use the `cacerts` or `cacertfile` options. _Which makes me wonder why AM keeps the PEM in the repo..._

## TL;DR

Maybe need a helper for a gateway that allows/requires client side SSL verification as an authentication mechanism. [List of such gateways supported by AM](https://github.com/activemerchant/active_merchant/search?utf8=%E2%9C%93&q=cert.pem&type=), we have encountered none so far.

**Comment by oyeb** [comment]
Closing this as we do not see any way to provide some common helpers to make connections. Vanilla HTTPoison` is best.
